OvmfPkg/IgvmSecureBootDxe/IgvmSecureBootDxe.c

```diff
@@ -0,0 +1,253 @@
+}
+
+BOOLEAN
+EFIAPI
+IgvmSecureBootIsConfigured (
+  VOID
+  )
+{
+  EFI_STATUS  Status;
+  BOOLEAN     Result;
+  UINTN       Size;
+
+  Status = gRT->GetVariable (
+                  L"PK",
+                  &gEfiGlobalVariableGuid,
+                  NULL,
+                  &Size,
+                  NULL
+                  );
+
+  if (Status == EFI_NOT_FOUND) {
+    Result = FALSE;
+  } else {
+    Result = TRUE;
+  }
+
+  DEBUG ((DEBUG_INFO, "%a: Result=%a\n", __func__, Result ? "Yes" : "No"));
+  return Result;
+}
+
+EFI_STATUS
+EFIAPI
+IgvmSecureBootDxeEntrypoint (
+  IN EFI_HANDLE        ImageHandle,
+  IN EFI_SYSTEM_TABLE  *SystemTable
+  )
+{
+  EFI_STATUS  Status;
+  UINT64      CcAttr;
+
+  CcAttr = PcdGet64 (PcdConfidentialComputingGuestAttr);
+
+  if (!CcAttr) {
+    /*
+     * The design idea is that the pages containing the HOBs and the EFI
+     * security databases are part of the launch measurement of a confidential
+     * VM.  This way the launch measurement will not only proof the guest runs
+     * the the firmware code it is supposed to run, but also proof secure boot
+     * is configured the way it is supposed to be.
+     *
+     * Any code loaded by the firmware will be subject to standard secure boot
+     * verification:
+     *  - Option one is to use using code signing and add the x509 vertificates
+     *    to 'db' (and revocation list to 'dbx'), which is the typical setup in
+     *    non-confidential VMs.
+     *  - More strict option two is to only allow specific efi binaries by
+     *    adding the sha256 authenticode hashes to 'db'.
+     *
+     * This workflow can only work in confidential VMs, so error out otherwise.
+     */
+    DEBUG ((DEBUG_ERROR, "%a: ERROR: not running in Confidential VM\n", __func__));
+    return EFI_UNSUPPORTED;
+  }
+
+  if (IgvmSecureBootIsConfigured ()) {
+    DEBUG ((DEBUG_ERROR, "%a: secure boot is already configured.\n", __func__));
+    return EFI_ALREADY_STARTED;
+  }
+
+  Status = IgvmSecureBootFindBlobs ();
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = IgvmSecureBootSetVariable (L"dbx", &gEfiImageSecurityDatabaseGuid, mDbx);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = IgvmSecureBootSetVariable (L"db", &gEfiImageSecurityDatabaseGuid, mDb);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = IgvmSecureBootSetVariable (L"KEK", &gEfiGlobalVariableGuid, mKEK);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // allow setting a platform key (PK) which is not self-signed
+  Status = IgvmSecureBootCustomMode (TRUE);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = IgvmSecureBootSetVariable (L"PK", &gEfiGlobalVariableGuid, mPK);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  Status = IgvmSecureBootCustomMode (FALSE);
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  DEBUG ((DEBUG_INFO, "%a: secure boot setup complete\n", __func__));
+  return EFI_SUCCESS;

```

Comment:
You could return `EFI_REQUEST_UNLOAD_IMAGE` so that the driver is unloaded as soon as it has done its job.